### ISSUE TITLE
Refactor interoperability

### ISSUE THREAD

**Issue by Simpag** [team-decent/decent-bench#273]
Currently the interoperability layer is a bit dense and hard to extend. My suggestion is to turn the interoperability layer into a class based system. I would define an abstract class containing all the methods which we currently support in the interoperability layer. Then each framework can implement all of these operations and register to the main handler. Something like the following:

```python
from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Any

import numpy as np

from decent_bench.utils.array import Array
from decent_bench.utils.types import SupportedArrayTypes, SupportedFrameworks

if TYPE_CHECKING:
    import torch

try:
    import torch

    TORCH_AVAILABLE = True
except ImportError:
    TORCH_AVAILABLE = False


class Interop(ABC):
    @abstractmethod
    def sum(self, array: Any, dim: int | tuple[int, ...] | None = None, keepdims: bool = False) -> Any:
        """Sum of array elements over a given axis."""


class NumpyIterop(Interop):
    def sum(
        self,
        array: np.ndarray[np.float64],
        dim: int | tuple[int, ...] | None = None,
        keepdims: bool = False,
    ) -> np.ndarray[np.float64]:
        res: np.ndarray[np.float64] = np.sum(array, axis=dim, keepdims=keepdims)
        return res


class TorchInterop(Interop):
    def sum(
        self,
        array: torch.Tensor,
        dim: int | tuple[int, ...] | None = None,
        keepdims: bool = False,
    ) -> torch.Tensor:  # Replace with actual torch.Tensor type when available
        return torch.sum(array, dim=dim, keepdim=keepdims)


class Interoperability(Interop):
    def __init__(self) -> None:
        super().__init__()
        self.backends: dict[SupportedFrameworks, Interop] = {}
        self.types: dict[SupportedFrameworks, tuple[type, ...]] = {}

    def register_backend(
        self, framework: SupportedFrameworks, backend: Interop, backend_types: tuple[type, ...]
    ) -> None:
        self.backends[framework] = backend
        self.types[framework] = backend_types

    def framework_of_array(self, array: SupportedArrayTypes) -> SupportedFrameworks:
        value = array.value if isinstance(array, Array) else array

        for framework, types in self.types.items():
            if isinstance(value, types):
                return framework

        raise TypeError(f"Unsupported framework type: {type(value)}")

    def _return_array(self, array: SupportedArrayTypes) -> Array:
        if not TYPE_CHECKING:
            return array

        return Array(array)

    def sum(
        self,
        array: Array,
        dim: int | tuple[int, ...] | None = None,
        keepdims: bool = False,
    ) -> Array:
        value = array.value if isinstance(array, Array) else array
        framework = self.framework_of_array(value)
        backend = self.backends[framework]
        result = backend.sum(value, dim=dim, keepdims=keepdims)
        return self._return_array(result)


# Create global instance and register backends
interoperability = Interoperability()

interoperability.register_backend(SupportedFrameworks.NUMPY, NumpyIterop(), (np.ndarray,))

if TORCH_AVAILABLE:
    interoperability.register_backend(SupportedFrameworks.PYTORCH, TorchInterop(), (torch.Tensor,))

```

There are still some parts left like the `SupportedFrameworks` and `SupportedDevices` which could be improved. Please let me know what you think and if you have any ideas.

**Comment by Simpag** [comment]
This refactoring should include a new `set_backend` method which is used to globally set the framework and device used during benchmark execution. This new `set_backend` method should be called on cost function initialization. The first call to `set_backend` should set the global backend and subsequent calls should be ignored if the same `framework/device` is passed and raise an error if a different one is passed. This makes sure that only one framework and device can be used during benchmark execution. How this is implemented in the cost function is up to whoever completes this issue but one idea is to use `__init_subclass__`.

**Comment by Simpag** [comment]
Might want to change how the `to_FRAMEWORK` methods work. There is no real reason to support any conversion between frameworks besides numpy. I can't see a use case where a user would convert a Tensorflow tensor to a PyTorch tensor.

**Comment by nicola-bastianello** [comment]
### some thoughts

I've been thinking, and I believe it would be better to explicitly ask the user to call `set_backend`. trying to infer what the framework and device are at runtime (e.g. from the cost definition) might be hacky. and if a user defines `b=iop.zeros(shape)` as e.g. data to be used in costs, at the time of that call the backend is not set, so we would need to still ask the user to write `b=iop.zeros(framework, device, shape)`, or to go with numpy as default

but I think it's better if the backend choice is intentional by the user, so that the user knows what it is without potential surprises if the backend is inferred/guessed. to make it as user friendly as possible, we have `def set_backend(framework: SupportedFrameworks, device: SupportedDevices): ...` like we do now, so just two strings are needed

### implementation idea

with this choice in mind, I think the following implementation would be nice:

1. as proposed by Simon, we have

```python
class _Backend(ABC):
    @abstractmethod
    def sum(self, array: Array, axis: int | tuple[int, ...] | None = None, keepdims: bool = False) -> Array:
```

defining the API, and then for each backend we subclass and implement the details

```python
class _NumpyBackend(_Backend)
    def sum(self, array: Array, axis: int | tuple[int, ...] | None = None, keepdims: bool = False) -> Array:
        res: np.ndarray[np.float64] = np.sum(array, axis=dim, keepdims=keepdims)
        return Array(res)
```

2. we then define `set_backend` to define the backend at runtime based on the configuration given by the user:
```python
_backend = None

def set_backend(framework, device, **kwargs):  # kwargs if any more config is needed for a backend
    global _backend
    if _backend is None:
        if framework == "numpy":
            _backend = _NumpyBackend()
        #  and all other supported backends, with raise at the end for unsupported
```

this sets the backend as global once the user has decided what it is, which they should do at the very start of their script
3. then, we need to expose the functionality in the API; to do this, we define

```python
def get_backend():
    if _backend is None:
        raise RuntimeError("Backend has not been defined")
    return _backend
```

and then expose all the same functions that we currently have (as functions, not as methods of a _Backend subclass):

```python
def sum(self, array: Array, axis: int | tuple[int, ...] | None = None, keepdims: bool = False) -> Array:
    return get_backend().sum(...)
```

this way, the API still looks like a collection of functions (like in numpy and the other frameworks), without asking the user to wrap their head around the `_Backend` class
4. and we support conversion from other framework to numpy, but only that

### other details

1. exposing the functions instead of `_Backend` is, in my opinion, cleaner and more intuitive for users. but it does lead to a bit of duplication, since the API definitions are given both in `_Backend`'s abstractmethods and in the interop functions; and it might be more difficult to maintain. asking the AI, it looks like we might solve this by `functools.wraps`. this is what I got from the AI, seems to make sense

```python
from functools import wraps
import inspect
from typing import Any, Callable

# create public functions as wrappers of the backend methods, with same signature/annotation/docs
def _create_wrapper(method_name: str) -> Callable:
    """Factory for one wrapper function."""
    abstract_method = getattr(_Backend, method_name)
    
    @wraps(abstract_method)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        backend = get_backend()
        return getattr(backend, method_name)(*args, **kwargs)
    
    # preserve metadata
    wrapper.__signature__ = inspect.signature(abstract_method)
    wrapper.__annotations__ = abstract_method.__annotations__
    wrapper.__doc__ = abstract_method.__doc__
    
    return wrapper

# automatic generation of the API
def _expose_all_methods(abc_cls: type[_Backend]) -> None:
    """Auto-populate module globals with wrappers for all abstract methods."""
    for name, member in abc_cls.__dict__.items():
        if callable(member) and getattr(member, '__isabstractmethod__', False):
            globals()[name] = _create_wrapper(name)

# at import time, the API is generated
_expose_all_methods(_Backend)

# expose the generated functions
__all__ = []  # Auto-populated below
__all__ = [name for name in globals() if name in _Backend.__dict__ and callable(getattr(_Backend, name, None)) and getattr(getattr(_Backend, name), '__isabstractmethod__', False)]
```

2. when writing the backends, we could provide `@to_array` for convenience to avoid explicitly wrapping the result of each method in Array. I'm not sure if mypy allows this, and it doesn't simplify syntax all that much

**Comment by nicola-bastianello** [comment]
when working on #287 , I realized that the logger is started only when the benchmark is called, so I had to write a hacky solution to start the logger inside `create_X_problem` if it is not already started. later, benchmark rewrites its settings from inside `benchmark` based on `log_level` argument

as I see it, the following config parameters should all be defined at the very start of any script:

- the backend (framework and device)
- the logger level
- the random seed
- and maybe: the checkpoint manager folder

so, we could collect all of these setup parameters in a single function

```python
def config(framework: SupportedFrameworks,
                 device: SupportedDevices,
                 seed: int | None = None,
                 log_level: int = logging.INFO,
                 storage_dir: str | Path = None  # maybe
      ) -> StorageManager:  # or None if StorageManager is defined separately
```

I think this is a very user friendly solution, and it solves quite a lot of headaches of trying to work around the lack of a config. later down the line we could also allow passing a config file into `config` which is parsed by [hydra](https://hydra.cc/docs/intro/); and in turn the config file could be created through a GUI

**Comment by Simpag** [comment]
The logger module should be updated so that the logger is started if the LOGGER is imported. I've noticed this issue too but havent had the time to fix it. 

Regarding the set_backend. This would still require some form of compatibility checks so that a user dont set a backend that isnt supported for the cost function they chose. I see two options, either we have a set_backend method with compliance checks during runtime (before algorithm execution) so users can use the iop methods before the cost function is created or we keep the current framework + device parameters but when they are set to none we try to infer which backend is being used, if no cost function is created yet it will throw an error and users must specify the backend. set_backend would not be a requirement, only if users wants to use iop before cost creation. However, I dont think users will need to use the iop methods before cost creation as all cost methods expect data in its native framework?

**Comment by nicola-bastianello** [comment]
yes, I agree that users who know the backend will likely not use iop for cost creation. but consider a user who doesn't know the syntax to all the backends, but still wants to experiment with them: they learn the iop syntax and use those functions to create the costs

also, it might be that some unforeseen order of operations would call iop before costs have been defined

so I think I would prefer to rule out inference from costs definition; let's go for set_backend with compliance checks, and actually I quite like the idea of wrapping set_backend insider a `config` method. then we add to all costs a compliance check when they are initialized, using `get_backend` (which raises if the user didn't call `config` or `set_backend`)

**Comment by nicola-bastianello** [comment]
I agree that the logger should be initialized at import; but the problem of defining the log_level still stands, so doing that in `config` makes a lot of sense
